client/peerset/src/peer_store.rs

```diff
@@ -0,0 +1,373 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use libp2p::PeerId;
+use log::trace;
+use partial_sort::PartialSort;
+use std::{
+	cmp::{Ord, Ordering, PartialOrd},
+	collections::{hash_map::Entry, HashMap, HashSet},
+	fmt::Debug,
+	sync::{Arc, Mutex},
+	time::{Duration, Instant},
+};
+use wasm_timer::Delay;
+
+use crate::{ReputationChange, LOG_TARGET};
+
+/// We don't accept nodes whose reputation is under this value.
+pub const BANNED_THRESHOLD: i32 = 82 * (i32::MIN / 100);
+/// Reputation change for a node when we get disconnected from it.
+const DISCONNECT_REPUTATION_CHANGE: i32 = -256;
+/// Relative decrement of a reputation value that is applied every second. I.e., for inverse
+/// decrement of 50 we decrease absolute value of the reputation by 1/50. This corresponds to a
+/// factor of `k = 0.98`. It takes ~ `ln(0.5) / ln(k)` seconds to reduce the reputation by half,
+/// or 34.3 seconds for the values above. In this setup the maximum allowed absolute value of
+/// `i32::MAX` becomes 0 in ~1100 seconds (actually less due to integer arithmetic).
+const INVERSE_DECREMENT: i32 = 50;
+/// Amount of time between the moment we last updated the [`PeerStore`] entry and the moment we
+/// remove it, once the reputation value reaches 0.
+const FORGET_AFTER: Duration = Duration::from_secs(3600);
+
+pub trait PeerReputationProvider: Debug + Send {
+	/// Check whether the peer is banned.
+	fn is_banned(&self, peer_id: &PeerId) -> bool;
+
+	/// Report the peer disconnect for reputation adjustement.
+	fn report_disconnect(&mut self, peer_id: PeerId);
+
+	/// Adjust peer reputation value.
+	fn report_peer(&mut self, peer_id: PeerId, change: ReputationChange);
+
+	/// Get peer reputation value.
+	fn peer_reputation(&self, peer_id: &PeerId) -> i32;
+
+	/// Get the candidates for initiating outgoing connections.
+	fn outgoing_candidates(&self, count: usize, ignored: HashSet<&PeerId>) -> Vec<PeerId>;
+}
+
+#[derive(Debug, Clone)]
+pub struct PeerStoreHandle {
+	inner: Arc<Mutex<PeerStoreInner>>,
+}
+
+impl PeerReputationProvider for PeerStoreHandle {
+	fn is_banned(&self, peer_id: &PeerId) -> bool {
+		self.inner.lock().unwrap().is_banned(peer_id)
+	}
+
+	fn report_disconnect(&mut self, peer_id: PeerId) {
+		self.inner.lock().unwrap().report_disconnect(peer_id)
+	}
+
+	fn report_peer(&mut self, peer_id: PeerId, change: ReputationChange) {
+		self.inner.lock().unwrap().report_peer(peer_id, change)
+	}
+
+	fn peer_reputation(&self, peer_id: &PeerId) -> i32 {
+		self.inner.lock().unwrap().peer_reputation(peer_id)
+	}
+
+	fn outgoing_candidates(&self, count: usize, ignored: HashSet<&PeerId>) -> Vec<PeerId> {
+		self.inner.lock().unwrap().outgoing_candidates(count, ignored)
+	}
+}
+
+impl PeerStoreHandle {
+	/// Get the number of known peers.
+	pub fn num_known_peers(&self) -> usize {
+		// FIXME: how do we use this info? May be better ask for numbers from protocol controllers?
```

Comment:
I see. Could we get that from `NetworkWorker` but calculating the number of peers we have a notification protocol open? Maybe something in `Protocol` (or in the future `Notifications`) could track that information and it wouldn't have to be split across `ProtocolController` instances.

Reply:
Currently we can't get this info by calculating the number of peers we have a notification protocol open, because this number also include discovered, but not connected peers. In practice, I think, the metric reported by `PeerStore` shouldn't differ too much from the original number, because for peer to get forgotten it's reputation must be not updated for one hour, and that's unlikely to happen if we are communicating with the peer. Original `Peerset` number also excluded peers that were disconnected more than an hour ago and forgotten, so it wasn't exactly a number of "all discovered peers".

I'm going to add a code comment to inform a reader about this discrepancy.

Reply:
Done